[WGSL] Add reference types

tadeuzagallo opened the pull request:
#### 343b74e1ec191107d59e432dff63836078880f9c
<pre>
[WGSL] Add reference types
<a href="https://bugs.webkit.org/show_bug.cgi?id=251202">https://bugs.webkit.org/show_bug.cgi?id=251202</a>
&lt;rdar://problem/104689979&gt;

Reviewed by Myles C. Maxfield.

Remove the existing `TypeReference` node, which was a workaround to have a node
that points to another type node, but it&apos;s now no longer necessary.

Replace it with a `ReferenceType` node. The similar name is a bit confusing,
but this actually refers to a reference to type, i.e. `T&amp;` in source.

The node still isn&apos;t used in this patch, but another patch will follow this one.

* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
(isType):
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::EntryPointRewriter::getResolvedType):
(WGSL::EntryPointRewriter::constructInputStruct):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):

Canonical link: <a href="https://commits.webkit.org/259474@main">https://commits.webkit.org/259474@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12c193b4957d5ba4c83f2c91a9f6da9861421898

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104867 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13947 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37762 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114136 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15077 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4871 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97195 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113160 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110627 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11651 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94658 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39169 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93517 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26276 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80833 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7291 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27636 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7387 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4225 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13443 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47186 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6535 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9176 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

djg reviewed (approved):
:shipit:

litherum commented on Source/WebGPU/WGSL/EntryPointRewriter.cpp:
```diff
@@ -141,12 +138,10 @@ void EntryPointRewriter::constructInputStruct()
             m_emptySourceSpan,
-            parameter.m_name,
-            parameterType.copyRef(),
+            WTFMove(parameter.m_name),
```
Moving from the parameter sounds scary, are you sure you want the parameter's name to be empty after this?

Maybe you want String(parameter.m_name)? Or just change the constructor to StructMember to accept a String& in addition to a String&&?

tadeuzagallo replied:
The variable naming is a bit confusing, but `parameter` here refers to a struct that I build as part of this pass only for this purpose, so I'm sure we won't need it after.

litherum reviewed (approved):
i'm a bit confused about this:

> this actually refers to a reference to type, i.e. `T&` in source.

WGSL doesn't have this kind of grammar; T& will never appear in WGSL.

Do you mean this from the spec? https://gpuweb.github.io/gpuweb/wgsl/#reference-type

tadeuzagallo committed:
[WGSL] Add reference types
https://bugs.webkit.org/show_bug.cgi?id=251202
<rdar://problem/104689979>

Reviewed by Myles C. Maxfield.

Remove the existing `TypeReference` node, which was a workaround to have a node
that points to another type node, but it's now no longer necessary.

Replace it with a `ReferenceType` node. The similar name is a bit confusing,
but this actually refers to a reference to type, i.e. `T&` in source.

The node still isn't used in this patch, but another patch will follow this one.

* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
(isType):
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::EntryPointRewriter::getResolvedType):
(WGSL::EntryPointRewriter::constructInputStruct):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):

Canonical link: https://commits.webkit.org/259474@main
